### DISCUSSION TITLE
我想重构下服务端

### DISCUSSION THREAD

**Discussion by chjiyun** [walinejs/waline#1968] [General]
文档应该列出所有服务端api，以及接口的业务逻辑，这样就可以用其他语言重写（go、rust），node实在太占内存了，PM2起来就差不多200M，而且速度奇慢

---
The document should list all server-side APIs and the business logic of the interface, so that it can be rewritten in other languages ​​(go, rust), node is too much memory, PM2 is almost 200M, and the speed is extremely slow@@== ==I want to refactor the server

**Comment by lizheming** [comment]
https://waline.js.org/next/api/index.html 可以看下这个，是有所有的 API 接口文档的

---
https://waline.js.org/next/api/index.html You can take a look at this, there are all API interface documents

**Reply by chjiyun** [reply]
我在本地调试发现前端发出的请求没有前缀/api;
请问 登录用的jwt token吗，

---
I debugged locally and found that the request sent by the front end does not have the prefix /api;
May I ask the jwt token used for login?

**Reply by lizheming** [reply]
1. `/api` 是最近的一次改造，客户端需要等 v3 版本才会适配上，目前 v3 版本正在分支中，一些当前客户端无法享受到的最新修改可以看下这个 PR https://github.com/walinejs/waline/pull/1904 可以考虑等 v3 客户端出来做适配也是可以
2. 是 jwt token

---
1. `/api` is a recent transformation. The client needs to wait for the v3 version to adapt. Currently, the v3 version is being branched. Some of the latest changes that cannot be enjoyed by the current client can be found in this PR https:// github.com/walinejs/waline/pull/1904 You can consider waiting for the v3 client to come out for adaptation.
2. It is a jwt token

**Reply by JQiue** [reply]
注册时存到数据库的哈希密码是什么算法？

---
What is the algorithm for hashing passwords stored in the database during registration?

**Reply by lizheming** [reply]
@JQiue  Waline 支持用户自定义加密算法，默认使用的是一个通用的加密模块 phpass，这个加密模块有对应的各种语言的实现，可以搜一下

https://github.com/walinejs/waline/blob/main/packages/server/src/extend/controller.js#L43-L48
https://www.npmjs.com/package/phpass
https://crates.io/crates/phpass



---
@JQiue Waline supports user-defined encryption algorithms. The default encryption module phpass is a common encryption module. This encryption module has corresponding implementations in various languages. You can search for it.

https://github.com/walinejs/waline/blob/main/packages/server/src/extend/controller.js#L43-L48
https://www.npmjs.com/package/phpass
https://crates.io/crates/phpass

**Reply by JQiue** [reply]
https://github.com/walinejs/waline/blob/a3b0c477e9a1705622cbb35586601b9d1d2a5938/packages/server/src/controller/token.js#L64
这里的 jwt 调用的`this.config('jwtKey')`返回的是什么 key? 从哪里获取的？


---
https://github.com/walinejs/waline/blob/a3b0c477e9a1705622cbb35586601b9d1d2a5938/packages/server/src/controller/token.js#L64
What key is returned by `this.config('jwtKey')` called by jwt here? Where did it get it?

**Reply by lizheming** [reply]
@JQiue https://github.com/walinejs/waline/blob/a3b0c477e9a1705622cbb35586601b9d1d2a5938/packages/server/src/config/config.js#L53 可以看下这里，默认是用了一些数据库的密钥类字段作为 key，也可以通过 JWT_TOKEN 环境变量指定。

---
@JQiue https://github.com/walinejs/waline/blob/a3b0c477e9a1705622cbb35586601b9d1d2a5938/packages/server/src/config/config.js#L53 You can take a look here. By default, some database key fields are used as keys. It can also be specified through the JWT_TOKEN environment variable.

**Reply by JQiue** [reply]
已经看出来了，但我发现数据库表的 wl_Users.2fa 字段设计的是否不妥？这导致在其他语言建模出现了问题，因为不允许以数字开头

---
I have already seen it, but I found that the wl_Users.2fa field of the database table is not designed properly? This causes problems when modeling in other languages ​​because starting with a number

**Reply by lizheming** [reply]
@JQiue JS 场景没啥问题，相关数据库似乎也没禁止，如有冲突可以先忽略这个功能，2步认证感觉不是很常用

---
@JQiue There is no problem with the JS scenario, and the relevant database does not seem to prohibit it. If there is a conflict, you can ignore this function first. 2-step authentication does not seem to be very commonly used
